// @destination/prettier-plugin-twig/dist/index.js
"use strict";
const prettier_1 = require("prettier");
const parser_1 = require("./parser");
const printer_1 = require("./printer");
const languages = [
    {
        name: 'LiquidHTML',
        parsers: [parser_1.liquidHtmlLanguageName],
        extensions: ['.twig'],
        vscodeLanguageIds: ['twig', 'Twig'],
    },
];
const options = {
    liquidSingleQuote: {
        type: 'boolean',
        category: 'LIQUID',
        default: true,
        description: 'Use single quotes instead of double quotes in Liquid tags and objects.',
        since: '0.2.0',
    },
    embeddedSingleQuote: {
        type: 'boolean',
        category: 'LIQUID',
        default: true,
        description: 'Use single quotes instead of double quotes in embedded languages (JavaScript, CSS, TypeScript inside <script>, <style> or Liquid equivalent).',
        since: '0.4.0',
    },
    singleLineLinkTags: {
        type: 'boolean',
        category: 'HTML',
        default: false,
        description: 'Always print link tags on a single line to remove clutter',
        since: '0.1.0',
    },
    indentSchema: {
        type: 'boolean',
        category: 'LIQUID',
        default: false,
        description: 'Indent the contents of the {% schema %} tag',
        since: '0.1.0',
    },
};
const defaultOptions = {
    printWidth: 120,
};
const plugin2 = {
    languages,
    parsers: parser_1.parsers,
    printers: printer_1.printers2,
    options,
    defaultOptions,
};
const plugin3 = {
    languages,
    parsers: parser_1.parsers,
    printers: printer_1.printers3,
    options,
    defaultOptions,
};
const prettierMajor = prettier_1.version.split('.')[0];
module.exports = prettierMajor === '2' ? plugin2 : plugin3;
